src/verify.c

```diff
@@ -755,6 +758,43 @@ static enum swupd_code deal_with_extra_files(struct manifest *manifest, bool fix
 	return ret;
 }
 
+static struct list *sanitize_files(struct list *bundles_files, struct list *all_files)
+{
+	struct list *next;
+	struct list *sanitized_files = NULL;
+	struct file *file1, *file2;
+
+	all_files = list_head(all_files);
+	bundles_files = list_head(bundles_files);
+
+	/* all_files have the complete consolidated list of files
+	 * from all bundles installed in the system, so we can replace
+	 * each file from the bundles_files with the equivalent in the
+	 * all_files to make sure they have the correct status */
+	while (bundles_files) {
+		next = bundles_files->next;
+		file1 = bundles_files->data;
+		/* bundles_list is a subset of the all_files list
+		 * and both lists should be already sorted, so we
+		 * can continue the search for the next file where
+		 * we left off from the previous file */
+		for (; all_files; all_files = all_files->next) {
+			file2 = all_files->data;
+			if (strcmp(file2->filename, file1->filename) == 0) {
+				sanitized_files = list_append_data(sanitized_files, file2);
```

Comment:
In this function you are adding some elements of list 'all_files' to the sanitized_files list. By doing that you don't know when you are freeing the all_files list which element you should free and which you shouldn't. In current code you aren't cleaning any element, so we are having memory leaks.

One solution would be to duplicate all elements that you are appending to the sanitized_files. It's the easiest one, but not ideal. Instead, it looks like you aren't using all_files anywhere else after calling this function. So instead of creating a new list with the files that are 'sanitized' you can filter the all_files list removing the files that doesn't make sense to be there (and freeing the data). So we don't need to duplicate the list and we know who to free at the end.

Reply:
Well, actually the data from all_files aren't cleared too. Because this files are kept in bundles and are all freed over there. So there's no need to be concerned about memory leaks here. My comment was incorrect.
But I'm still concerned about this function being O(n^2) while it could be O(n).